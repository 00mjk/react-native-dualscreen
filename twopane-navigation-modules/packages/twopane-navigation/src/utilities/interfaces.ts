import React, { ReactElement } from 'react';
import { IHeader } from '../shared/screenStore/headerStore/header.interface';

export interface IPaneComponent {
  key: string;
  paneElement: ReactElement;
  header?: IHeader;
  screen?: paneType;
}

export interface IPaneState {
  screen: Array<IPaneComponent>;
}

export interface IEmptyAction {
  type: string;
  payload: {
  };
}

export enum paneType {
  ONE = 'ONE',
  TWO = 'TWO'
}

export interface ITwoPaneAppProps {
  /**
   * Default element for ONE screen mode
   */
  onePaneDefault: IPaneComponent;

  /**
   * Default element for TWO screen mode
   */
  twoPaneDefault: IPaneComponent;

  /**
   * AppContainer if using React Navigation
   */
  navigationContainer?: JSX.Element;

  /**
   * Items to render in app drawer. If empty, drawer will not be included.
   */
  drawerItems?: JSX.Element[];
}

export interface IPaneHeaderProps extends IHeader {
}


