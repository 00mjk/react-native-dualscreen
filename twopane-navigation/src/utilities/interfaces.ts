import React, { ReactElement } from 'react';
import { IHeader } from '../shared/screenStore/headerStore/header.interface';
import { StyleProp, ViewStyle, TextStyle, ImageStyle } from 'react-native';

export interface IPaneComponent extends IBasePaneComponent {
  pane: paneType;
  isExtended: boolean;
}

export type Style = StyleProp<ViewStyle>;
export type StyleText = StyleProp<TextStyle>;
export type StyleImage = StyleProp<ImageStyle>;

export interface IConfig {
  paneBody? : Style;
  paneHeader? : Style;
  paneHeaderText?: StyleText;
  paneHeaderIcon?: StyleImage;
}

//TODO ADD DEFAULT HEADER HEIGHT SIZE & BODY HEIGHT
export interface IConfigComponent {
  onePane?: IConfig;
  twoPane?: IConfig;
}

export interface IBasePaneComponent {
  key: string;
  paneElement: ReactElement;
  header?: IHeader;
}

export interface IEmptyAction {
  type: string;
  payload: {
  };
}

export enum paneType {
  ONE = 'ONE',
  TWO = 'TWO',
}

export interface ITwoPaneAppProps {
  /**
   * Default element for ONE screen mode
   */
  onePaneDefault: IBasePaneComponent;

  /**
   * Default element for TWO screen mode
   */
  twoPaneDefault: IBasePaneComponent;

  /**
   * override default values of pane components
   */
  config?: IConfigComponent

  /**
   * AppContainer if using React Navigation
   */
  navigationContainer?: JSX.Element;
}



