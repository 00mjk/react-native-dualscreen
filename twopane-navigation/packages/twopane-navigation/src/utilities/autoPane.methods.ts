import utility from './utility.methods';
import twoPane from '../twoPane/twoPaneStore/twoPane.methods';
import onePane from '../onePane/onePaneStore/onePane.methods';
import react, { ReactElement } from 'react';
import { IHeader } from '../Shared/screenStore/headerStore/header.interface';


/**
 * Automatically pushes element to the top of the stack based on screen size.
 */
const Add = (key: string, element: ReactElement, header?: IHeader, isMergeONE = false, isMergeTWO = false) => {
  if (utility.isTwoPane()) {
    twoPane.Add(key, element, header, isMergeTWO);
  } else {
    onePane.Add(key, element, header, isMergeONE);
  }
};

/**
 * Automatically pushes element to the top of stack or if the key is already in the stack,
    move that key to the top of the stack based on screen size
 */
const AddOrMoveToFront = (key: string, element: ReactElement, header?: IHeader, isMergeONE = false, isMergeTWO = false,) => {
  if (utility.isTwoPane()) {
    twoPane.AddOrMoveToFront(key, element, header, isMergeTWO);
  } else {
    onePane.AddOrMoveToFront(key, element, header, isMergeONE);
  }
};

/**
 * Automatically pushes element to the top of stack or if the key is already in the  TWO screen stack,
    move that key to the top of the TWO screen stack based on screen size
 */
const AddOrMoveToFrontTWO = (key: string, element: ReactElement, header?: IHeader, isMergeONE = false, isMergeTWO = false,) => {
  if (utility.isTwoPane()) {
    twoPane.AddOrMoveToFront(key, element, header, isMergeTWO);
  } else {
    onePane.Add(key, element, header, isMergeONE);
  }
};

/**
 * Automatically pushes element to the top of stack or if the key is already in the ONE screen stack,
    move that key to the top of the ONE screen stack based on screen size
 */
const AddOrMoveToFrontONE = (key: string, element: ReactElement, header?: IHeader, isMergeONE = false, isMergeTWO = false,) => {
  if (utility.isTwoPane()) {
    twoPane.Add(key, element, header, isMergeTWO);
  } else {
    onePane.AddOrMoveToFront(key, element, header, isMergeONE);
  }
};

/**
 * Automatically Removes all elements of the stack and returns the base element of the screen stack based on screen size
 */
const BackToHome = () => {
  if (utility.isTwoPane()) {
    twoPane.BackToHome();
  } else {
    onePane.BackToHome();
  }
};


/**
 * Automatically go back one element in the stack based on screen size
 */
const GoBack = () => {
  if (utility.isTwoPane()) {
    twoPane.GoBack();
  } else {
    onePane.GoBack();
  }
};

/**
 * Automatically replace element in the stack based on screen size
 */
const ReplaceScreen = (key: string, twoPaneElement: React.ReactElement) => {
  if (utility.isTwoPane()) {
    twoPane.ReplaceScreen(key, twoPaneElement);
  } else {
    onePane.ReplaceScreen(key, twoPaneElement);
  }
};


/**
 * Automatically replace header in the stack based on screen size
 */
const ReplaceHeader = (key: string, header: IHeader) => {
  if (utility.isTwoPane()) {
    twoPane.ReplaceHeader(key, header);
  } else {
    onePane.ReplaceHeader(key, header);
  }
};



const _autoPaneFunctions = {
  Add,
  AddOrMoveToFront,
  AddOrMoveToFrontTWO,
  AddOrMoveToFrontONE,
  BackToHome,
  GoBack,
  ReplaceScreen,
  ReplaceHeader
};

type autoPaneFunctions = typeof _autoPaneFunctions;

const autoPane: autoPaneFunctions = _autoPaneFunctions;

export default autoPane;
