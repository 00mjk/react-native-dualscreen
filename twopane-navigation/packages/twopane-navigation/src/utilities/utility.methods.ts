import { useRef, useEffect } from 'react';
import { store } from '../appStore';

const isDualScreen = (): boolean => {
  return store.getState().utilityStoreReducer.isDualScreen
};

const _utilityFunctions = {
  isDualScreen,
};

type utilityFunctions = typeof _utilityFunctions;

const utility: utilityFunctions = _utilityFunctions;

export default utility;
