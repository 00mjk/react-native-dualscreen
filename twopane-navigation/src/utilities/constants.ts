export const hingeWidth: number = 33.6;