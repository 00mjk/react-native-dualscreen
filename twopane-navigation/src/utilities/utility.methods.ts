import React from 'react';
import { store } from '../appStore';

const isTwoPane = (): boolean => {
  return store.getState().utilityStoreReducer.isTwoPane
};

const _utilityFunctions = {
  isTwoPane,
};

type utilityFunctions = typeof _utilityFunctions;

const utility: utilityFunctions = _utilityFunctions;

export default utility;
