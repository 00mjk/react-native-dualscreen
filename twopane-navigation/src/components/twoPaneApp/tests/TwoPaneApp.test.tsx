import React, { Fragment } from 'react';
import DualApp from '../twoPaneApp';
import { IPaneComponent } from '../../../utilities/interfaces';
import { render, toJSON } from '@testing-library/react-native';

describe('DualApp Tests', () => {
    it('should render singleScreen', () => {
        // Arrange
        jest.useFakeTimers();
        
        const _onePaneDefault: IPaneComponent = {
            key: 'singleDefault',
            paneElement: <Fragment />,
            header: {
                title: 'singleScreenDefaultTitle'
            }
        };

        const _twoPaneDefault: IPaneComponent = {
            key: 'dualDefault',
            paneElement: <Fragment />,
            header: {
                title: 'onePaneDefaultTitle'
            }
        };

        // Act
        const { container } = render(
            <DualApp
                onePaneDefault={_onePaneDefault}
                twoPaneDefault={_twoPaneDefault}
            />
        );
        // TODO MOCK TEST
        // Assert
        expect(toJSON(container)).toMatchSnapshot();
    });
});
