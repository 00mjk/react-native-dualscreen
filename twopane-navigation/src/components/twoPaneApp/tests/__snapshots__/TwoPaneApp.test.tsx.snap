// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DualApp Tests should render singleScreen 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Array [
          Object {
            "bottom": 0,
            "color": "black",
            "flex": 1,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": 720,
            "left": 0,
            "width": 540,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "height": "10%",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                },
                Array [
                  Object {
                    "backgroundColor": "#212121",
                    "paddingHorizontal": 25,
                  },
                  undefined,
                ],
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#F2F2F2",
                  "fontSize": 20,
                  "fontWeight": "500",
                  "marginLeft": 6,
                }
              }
            >
              singleScreenDefaultTitle
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": "85%",
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "bottom": 0,
            "color": "black",
            "flex": 1,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": 720,
            "left": 573.6,
            "width": 540,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "height": "10%",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                },
                Array [
                  Object {
                    "backgroundColor": "#212121",
                    "paddingHorizontal": 25,
                  },
                  undefined,
                ],
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#F2F2F2",
                  "fontSize": 20,
                  "fontWeight": "500",
                  "marginLeft": 6,
                }
              }
            >
              onePaneDefaultTitle
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": "85%",
          }
        }
      />
    </View>
  </View>
</View>
`;
